rclcpp/include/rclcpp/mapped_ring_buffer.hpp

```diff
@@ -101,57 +102,98 @@ class MappedRingBuffer : public MappedRingBufferBase
    * \param value if the key is found, the value is stored in this parameter
    */
   void
-  get_copy_at_key(uint64_t key, ElemUniquePtr & value)
+  get(uint64_t key, ElemUniquePtr & value)
   {
     std::lock_guard<std::mutex> lock(data_mutex_);
     auto it = get_iterator_of_key(key);
     value = nullptr;
     if (it != elements_.end() && it->in_use) {
-      auto ptr = ElemAllocTraits::allocate(*allocator_.get(), 1);
-      ElemAllocTraits::construct(*allocator_.get(), ptr, *it->value);
-      value = ElemUniquePtr(ptr);
+      T * raw_ptr;
+      ElemDeleter * deleter;
+      if (it->unique_value) {
+        raw_ptr = it->unique_value.get();
+        deleter = &it->unique_value.get_deleter();
+      } else {
+        raw_ptr = const_cast<T *>(it->shared_value.get());
```

Comment:
maybe I haven't understood it but it feels the same `raw_ptr` will be deleted twice because it is owned by two different smart pointers

Reply:
Yes, the `raw_ptr` is only used to construct a new unique pointer, although this is probably a little confusing. I will try to do it cleaner.